### PR TITLE
add split test

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

File: server/cluster.go [hunk 1]
```diff
@@ -148,6 +148,10 @@ func encodeRegionSearchKey(endKey []byte) string {
 	return string(append([]byte{'z'}, endKey...))
 }
 
+func encodeRegionStartKey(startKey []byte) string {
```
Comment (qiuyesuifeng) [inline server/cluster.go]: Can we reuse encodeRegionSearchKey function?
Change `endKey` variable to `key`?

Comment (disksing) [inline server/cluster.go]: @qiuyesuifeng No. Separate them guarantees that 'encoded startKey nil' < 'encoded endKey nil'. StartKey and EndKey will be compared at L262.

File: server/cluster_worker.go [hunk 2]
```diff
@@ -233,24 +253,14 @@ func (c *raftCluster) maybeSplit(request *pdpb.RegionHeartbeatRequest, reqRegion
...
-
-	regionEncStartKey := encodeRegionSearchKey(region.GetStartKey())
+	regionEncStartKey := encodeRegionStartKey(region.GetStartKey())
 	regionEncEndKey := encodeRegionSearchKey(region.GetEndKey())
-	reqRegionEncStartKey := encodeRegionSearchKey(reqRegion.GetStartKey())
-	reqRegionEncEndKey := encodeRegionSearchKey(reqRegion.GetEndKey())
+
 	if reqRegionEncStartKey == regionEncStartKey &&
 		reqRegionEncEndKey == regionEncEndKey {
 		// Seems there is something wrong? Do nothing.
```
Comment (siddontang) [inline server/cluster_worker.go]: Wrong comment here? 
Here seems no split?

Comment (disksing) [inline server/cluster_worker.go]: @siddontang In that case, the 'epoch.Version' will be unchanged and this func will return at L253 above.

Comment (siddontang) [inline server/cluster_worker.go]: Yes, so I think this is not "something wrong", this is correct.

Comment (qiuyesuifeng) [inline server/cluster_worker.go]: Yes, seems can be updated.

File: server/cluster.go [hunk 3]
```diff
@@ -355,15 +359,15 @@ func (c *raftCluster) GetRegion(regionKey []byte) (*metapb.Region, error) {
 		return nil, errors.Trace(err)
 	}
 	if !ok {
-		return nil, errors.Errorf("we must find a region for %q but fail, a serious bug", regionKey)
+		return nil, nil
```
Comment (qiuyesuifeng) [inline server/cluster.go]: Should we always get region?
Can u show an example for this change?

Comment (siddontang) [inline server/cluster.go]: we can't always get region.

Comment (siddontang) [inline server/cluster.go]: E,g, for first region ["", +inf), split to ["", b), [b, +inf), when the left region reports first, we have only ["", b) one region in pd now, so using any key >= b will get no region.

Comment (qiuyesuifeng) [inline server/cluster.go]: Gotcha.

File: server/cluster_worker_test.go [hunk 4]
```diff
@@ -227,6 +226,129 @@ func (s *testClusterWorkerSuite) checkChangePeerRes(c *C, res *pdpb.RegionHeartb
...
+
+	resp = s.heartbeatRegion(c, conn, 0, r3, leaderPeer3)
+	c.Assert(resp, IsNil)
+
+	r, err = cluster.GetRegion([]byte("z"))
+	c.Assert(err, IsNil)
+	c.Assert(r, DeepEquals, r3)
+
+	// [m, q) is missing before r2's heartbeat.
+	r, err = cluster.GetRegion([]byte("n"))
```
Comment (qiuyesuifeng) [inline server/cluster_worker_test.go]: I think we should also check `/1/raft/r/r1 -> encoded_region_key` value.

File: server/cluster_worker_test.go [hunk 5]
```diff
@@ -227,6 +226,129 @@ func (s *testClusterWorkerSuite) checkChangePeerRes(c *C, res *pdpb.RegionHeartb
...
+			Region: region,
+		},
+	}
+	sendRequest(c, conn, msgID, req)
+	_, resp := recvResponse(c, conn)
+	c.Assert(resp.GetCmdType(), Equals, pdpb.CommandType_RegionHeartbeat)
+	return resp.GetRegionHeartbeat().GetChangePeer()
+}
+
+func (s *testClusterWorkerSuite) TestSplit(c *C) {
```
Comment (qiuyesuifeng) [inline server/cluster_worker_test.go]: @siddontang  Should we simulate a test with heartbeat both with split and peer change update?

**Comment by disksing** [comment]
@qiuyesuifeng PTAL

File: server/cluster_worker_test.go (hunk 5 above)
Comment (siddontang) [inline server/cluster_worker_test.go]: yep, but the probability is little for this case, I think we can do it later, can fire an issue first.

**Comment by siddontang** [comment]
LGTM

File: server/cluster_worker_test.go (hunk 4 above)
Comment (qiuyesuifeng) [inline server/cluster_worker_test.go]: @disksing  How about this?

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**